### PR TITLE
Cross-language invocation Part 1: Java calling Python functions and actors

### PR DESCRIPTION
**Pull request by raulchen**
## What do these changes do?

Implements Java calling Python functions and actors.

Usage:
- Call a Python function: 
```java
RayObject res = Ray.callPy("module_name", "function_name", arg1, arg2, ...);
```
- Call a Python actor:
```java
RayPyActor actor = Ray.createPyActor("module_name", "ClassName", arg1, arg2, ...);
RayObject res = Ray.callPy(actor, "function_name", arg1, arg2);
```

Note, currently arguments can only be raw binaries. And when putting a binary in object store, we use special metadata to indicate that the object is raw binary and doesn't need serialization/deserialization.

Not implemented in this PR:
- Python calling Java.
- Passing actor handle as parameter.

## Related issue number

#2576

### CODE REVIEW

**Comment by pcmoritz** [comment]
If we got the eternal bug https://issues.apache.org/jira/browse/ARROW-1692 resolved, we could pass data natively between Java arrow and python :D

File: java/runtime/src/main/java/org/ray/runtime/AbstractRayRuntime.java [hunk 1]
```diff
@@ -260,16 +263,69 @@ public RayObject call(RayFunc func, RayActor<?> actor, Object[] args) {
...
+        pyActor.getClassName(), functionName);
+    RayPyActorImpl actorImpl = (RayPyActorImpl) pyActor;
+    TaskSpec spec;
+    synchronized (pyActor) {
+      spec = createTaskSpec(null, desc, actorImpl, args, false, null);
+      spec.getExecutionDependencies().add(actorImpl.getTaskCursor());
+      actorImpl.setTaskCursor(spec.returnIds[1]);
+      actorImpl.clearNewActorHandles();
+    }
+    rayletClient.submitTask(spec);
```
Comment (ericl) [inline java/runtime/src/main/java/org/ray/runtime/AbstractRayRuntime.java]: Nice.

File: java/runtime/src/main/java/org/ray/runtime/AbstractRayRuntime.java [hunk 2]
```diff
@@ -260,16 +263,69 @@ public RayObject call(RayFunc func, RayActor<?> actor, Object[] args) {
...
+    PyFunctionDescriptor desc = new PyFunctionDescriptor(moduleName, className, "__init__");
+    TaskSpec spec = createTaskSpec(null, desc, RayPyActorImpl.NIL, args, true, options);
+    RayPyActorImpl actor = new RayPyActorImpl(spec.actorCreationId, moduleName, className);
+    actor.increaseTaskCounter();
+    actor.setTaskCursor(spec.returnIds[0]);
+    rayletClient.submitTask(spec);
+    return actor;
+  }
+
+
```
Comment (ericl) [inline java/runtime/src/main/java/org/ray/runtime/AbstractRayRuntime.java]: Too many newlines?

File: java/pom.xml [hunk 3]
```diff
@@ -130,6 +130,11 @@
         <artifactId>jaxb-impl</artifactId>
         <version>${jaxb.version}</version>
       </dependency>
+      <dependency>
+        <groupId>com.google.code.gson</groupId>
+        <artifactId>gson</artifactId>
+        <version>2.8.5</version>
+      </dependency>
```
Comment (ericl) [inline java/pom.xml]: What's gson for?

**Review by ericl — CHANGES_REQUESTED** [review]
Left a few comments.

File: java/api/src/main/java/org/ray/api/RayCall.java [hunk 4]
```diff
@@ -2312,4 +2310,143 @@
...
+    return Ray.internal().createPyActor(moduleName, className, args, options);
+  }
+  public static RayPyActor createPyActor(String moduleName, String className, Object obj0, Object obj1, Object obj2, Object obj3, Object obj4, Object obj5) {
+    Object[] args = new Object[]{obj0, obj1, obj2, obj3, obj4, obj5};
+    return Ray.internal().createPyActor(moduleName, className, args, null);
+  }
+  public static RayPyActor createPyActor(String moduleName, String className, Object obj0, Object obj1, Object obj2, Object obj3, Object obj4, Object obj5, ActorCreationOptions options) {
+    Object[] args = new Object[]{obj0, obj1, obj2, obj3, obj4, obj5};
+    return Ray.internal().createPyActor(moduleName, className, args, options);
+  }
```
[the remark this answers is not in the document]
Comment (raulchen) [reply java/api/src/main/java/org/ray/api/RayCall.java]: yeah, varargs would look nicer. But I wanted to make the order aligned with the regular `createActor`.

File: java/runtime/src/main/java/org/ray/runtime/RayPyActorImpl.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+  public String getModuleName() {
+    return moduleName;
+  }
+
+  @Override
+  public String getClassName() {
+    return className;
+  }
+
+  public RayPyActorImpl fork(boolean random) {
```
[the remark this answers is not in the document]
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/RayPyActorImpl.java]: Oops, that's an internal hack. I'll remove it.

File: java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java [hunk 6]
```diff
@@ -250,12 +251,29 @@ private static ByteBuffer convertTaskSpecToFlatbuffer(TaskSpec task) {
...
-    int[] functionDescriptorOffsets = new int[]{
-        fbb.createString(task.functionDescriptor.className),
-        fbb.createString(task.functionDescriptor.name),
-        fbb.createString(task.functionDescriptor.typeDescriptor)
-    };
-    int functionDescriptorOffset = fbb.createVectorOfTables(functionDescriptorOffsets);
+    int language;
+    int functionDescriptorOffset;
+
+    if (task.functionDescriptor != null) {
```
[the remark this answers is not in the document]
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/raylet/RayletClientImpl.java]: sounds good.

File: java/pom.xml (hunk 3 above)
Comment (raulchen) [reply java/pom.xml]: It's actually needed for the next PR (serializing handle with json and pass it to Python), forgot to remove it.

**Comment by ericl** [comment]
@raulchen any follow-up here? Looks like there are some merge conflits now.

**Comment by raulchen** [comment]
@ericl I rebased the PR and addressed your comments.

**Comment by raulchen** [comment]
CI failure is related, but I cannot reproduce it on my machine. Will check tomorrow.

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/PyFunctionDescriptor.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package org.ray.runtime.functionmanager;
+
+/**
+ * Represents metadata of a Python function.
+ */
+public class PyFunctionDescriptor {
```
Comment (ericl) [inline java/runtime/src/main/java/org/ray/runtime/functionmanager/PyFunctionDescriptor.java]: For the sake of a little more type safety, how about having both this and FunctionDescriptor extend the same interface or base class?

**Review by ericl — APPROVED** [review]
Looks good -- just one comment on having a common base class for a bit more type safety.

It would also be nice to add an example of Python->Java calls, if there is a use case for that.

File: java/runtime/src/main/java/org/ray/runtime/functionmanager/PyFunctionDescriptor.java (hunk 7 above)
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/functionmanager/PyFunctionDescriptor.java]: Thanks, will change.

**Comment by raulchen** [comment]
@ericl Python calling Java will be implemented in next PR. Although it's already doable with low-level api, I'd like to create a higher-level API for convenience in next PR. I haven'd thought thoroughly about the API, it may be something like the following. Suggestions are welcome!

```
java_func = ray.experimental.java_task("package.name", "ClassName", "staticMethodName", "typeDescriptor")
java_func.remote(args)

java_actor = ray.experimental.create_java_actor("package.name", "ClassName")
java_actor.method("methodName", "typeDescriptor").remote(args)
```

**Comment by ericl** [comment]
Sounds good. I think the experimental syntax looks fine for an initial release; you could also try to wrap the actors to provide a bit more of a pythonic API.

**Pull request merged**